### PR TITLE
Ntp offset

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

File: ntp.go [hunk 1]
```diff
@@ -150,11 +150,11 @@ func Offset(host string) (int64, error) {
 
 // offset variable names based off the rfc:
 // https://tools.ietf.org/html/rfc2030
-func offset(t1, t2, t3, t4 ntpTime) (int64, error) {
+func offset(clientSend, serverReceive, serverTransmit, clientReceive ntpTime) (int64, error) {
```
Comment (beevik) [inline ntp.go]: Any reason why this is a separate helper function? Couldn't it be done directly in the Offset function?

File: ntp.go [hunk 2]
```diff
@@ -42,28 +42,8 @@ func (t ntpTime) UTC() time.Time {
...
-func (r ntpTime) add(a ntpTime) ntpTime {
-	var result ntpTime
-	result.Fraction = r.Fraction + a.Fraction
-	result.Seconds = r.Seconds + a.Seconds
-	if result.Fraction < a.Fraction {
-		result.Seconds += 1
-	}
-	return result
+func (t ntpTime) nsec() int64 {
+	return int64(t.Seconds) * 1e9 + int64(float64(t.Fraction) / frac * 1e9)
```
Comment (beevik) [inline ntp.go]: I'm not sure why you changed the calculation to do a conversion to and from floating point here.  What was wrong with the original method?

Comment (beevik) [inline ntp.go]: Here is a suggested version of this code that doesn't require any floating point operations:

```go
var (
	ntpEpoch = time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC)
)

func (t ntpTime) UTC() time.Time {
	nsec := uint64(t.Seconds)*1e9 + (uint64(t.Fraction) * 1e9 >> 32)
	return ntpEpoch.Add(time.Duration(nsec))
}

func toNtpTime(t time.Time) ntpTime {
	nsec := uint64(t.UTC().Sub(ntpEpoch))
	sec, frac := nsec/1e9, nsec%1e9
	return ntpTime{
		Seconds:  uint32(sec),
		Fraction: uint32(frac << 32 / 1e9),
	}
}
```

**Comment by beevik** [comment]
I see now that the "toNtpDate" change has gone through a bit of a back-and-forth with another commenter.  (I didn't notice it earlier because that PR was closed.)

Regarding the NTP time conversions, I'd suggest the following:
```go
var (
	ntpEpoch = time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC)
)

func (t ntpTime) UTC() time.Time {
	nsec := uint64(t.Seconds)*1e9 + (uint64(t.Fraction) * 1e9 >> 32)
	return ntpEpoch.Add(time.Duration(nsec))
}

func toNtpTime(t time.Time) ntpTime {
	nsec := uint64(t.UTC().Sub(ntpEpoch))
	sec, frac := nsec/1e9, nsec%1e9
	return ntpTime{
		Seconds:  uint32(sec),
		Fraction: uint32(frac << 32 / 1e9),
	}
}
```

This code should do what you want, and it doesn't require any floating point arithmetic.

File: ntp.go (hunk 1 above)
Comment (chbatey) [inline ntp.go]: So it can be tested without having to callout to a real ntp server

**Comment by chbatey** [comment]
I can update it to that. The only reason I went with floating point in the end was that the ntp reference implementation uses it.

**Comment by chbatey** [comment]
@beevik I have updated based on your comments and rebased with master after the other PR was merged

**Comment by beevik** [comment]
FYI, I unified the Query and Offset APIs.  You should use Query now.

**Comment by beevik** [comment]
Thanks for the pull request!

**Pull request merged**